### ISSUE TITLE
[Android] Setup failed when using avd capabilities

### ISSUE THREAD

**Issue by zethyscythe** [appium/appium#14347]
## The problem
I'm using capability avd and uiautomator2 for launching emulator from my local windows desktop and getting error : Error: Appium Settings app is not running after 5000ms. 
I saw the log and notice this command : 
```
[debug] [ADB] Running 'C:\Users\root\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s emulator-5554 shell am start -n io.appium.settings/.Settings -a android.intent.action.MAIN -c android.intent.category.LAUNCHER'
[debug] [ADB] Getting IDs of all 'io.appium.settings' processes
[debug] [ADB] Running 'C:\Users\root\AppData\Local\Android\Sdk\platform-tools\adb.exe -P 5037 -s emulator-5554 shell pgrep -f io\\.appium\\.settings'
```
Later i try to run this command manually and getting no results. However using the same capabilities for pre launched android emulator  this command showing result. Seems there are problems when launching the io.appium.setting activity.
```
adb -P 5037 -s emulator-5554 shell pgrep -f io\\.appium\\.settings
```

## Environment
Appium version -v1.17.1
Node.js version -v12.13.1
Npm or Yarn package manager: 6.14.2
Mobile platform/version under test: Android API 28
Real device or emulator/simulator: Android emulator
Appium CLI or Appium.app|exe: appium cli
## Details
Launch with uiautomator2

Link to Appium logs-https://gist.github.com/zethyscythe/f74dd3d729117a939bb58a1f8fb9476e

**Comment by shibupanda** [comment]
>[AndroidBootstrap] [BOOTSTRAP LOG] [debug] Using: UiSelector[INSTANCE=0, RESOURCE_ID=kr.co.kra.m.mycard3:id/lay_close]
[AndroidBootstrap] [BOOTSTRAP LOG] [debug] Returning result: {"status":7,"value":"No element found"}
[AndroidBootstrap] Received command result from bootstrap
[MJSONWP] Matched JSONWP error code 7 to NoSuchElementError
[W3C (a33ae738)] Encountered internal error running command: NoSuchElementError: An element could not be located on the page using the given search parameters.
[W3C (a33ae738)]     at AndroidDriver.findElOrEls (C:\Users\Akalanka Udesh\AppData\Local\Programs\Appium\resources\app\node_modules\appium\node_modules\appium-android-driver\lib\commands\find.js:75:11)
[W3C (a33ae738)]     at process._tickCallback (internal/process/next_tick.js:68:7)
[HTTP] <-- POST /wd/hub/session/a33ae738-eef3-45a7-8ce9-cbc16971e775/element 404 2212 ms 

@zethyscythe 
As per the logs there is no such error I found instead I found driver status 200 and after that it failed while finding elements.

**Comment by mykola-mokhnach** [comment]
Please attach the correct log.

Possible duplicate of https://github.com/appium/appium/issues/14341

**Comment by zethyscythe** [comment]
Sorry i forgot to change the link earlier as it was copy paste from other issue. I already edit the correct link in the detail

**Comment by mykola-mokhnach** [comment]
From the log it looks like the emulator startup is very slow on the machine, so it ignores the command to start Settings helpers, because the booting process has not been completed yet. Consider getting a faster machine to run the tests or use a prebooted emulator.

**Comment by zethyscythe** [comment]
From the logs the emulator is able to boot in 14 seconds,is it considered very slow?
Moreover is there any capabilities to delay the install/launch of io.appium.settings?

**Comment by mykola-mokhnach** [comment]
The trick here is emulator readiness detection. We could probably add a capability to delay the driver startup for a given amount of time after avd booting is completed.

**Comment by zethyscythe** [comment]
Yeah that's gonna works too. Will this capability available in future implementation then?

**Comment by mykola-mokhnach** [comment]
The fact that the boot animation is disabled could also be the case, since emulator startup is considered finished if no boot animation is shown anymore (`init.svc.bootanim` prop). The actual command line params disable that animation.

**Comment by zethyscythe** [comment]
Hi, you were right, After I removed `-no-boot-anim` the driver setup successfully. Seems the setting went failed because the adb detect emulator boot incorrectly. Simply remove this solved the problem.

**Comment by zethyscythe** [comment]
Also i found that since emulator 29.0.7 there is command line  option `-delay-adb` that can be used to make sure adb command package sent only after the emulator boot is completed.

This issue is resolved, thank you for your support @mykola-mokhnach .

**Comment by mykola-mokhnach** [comment]
> Also i found that since emulator 29.0.7 there is command line option -delay-adb that can be used to make sure adb command packets sent only after the emulator boot is completed.

Thanks for letting us know. I'l check if this option could be useful for us while detecting emulator boot

**Comment by mykola-mokhnach** [comment]
The option has been added to appium@beta

**Comment by Enrico204** [comment]
`-delay-adb` is causing some emulator images to hangs indefinitely. I have an Android 6 image (`system-images;android-23;google_apis;x86_64`) which hangs if started with `-delay-adb`. Android 9 image is working perfectly. I suggest to add an option to disable this flag.

**Comment by mykola-mokhnach** [comment]
@Enrico204 have you tried to report the issue to Google?

**Comment by Enrico204** [comment]
I will :-) However I think that this option should be customizable regardless.

**Comment by Enrico204** [comment]
I just found this issue in the Google Android Emulator issue tracking: https://issuetracker.google.com/issues/142533355

In short: comment 4 from a Google employee says that the `delay-adb` flag should be used only on **API 28 and later** (API <= 27 images don't send the "boot completed" signal required).

So the correct implementation is to enable this flag only on Android versions >= 9

**Comment by mykola-mokhnach** [comment]
This makes sense. I'll add a patch. Thanks for the update @Enrico204
